substrate-archive/src/tasks.rs

```diff
@@ -90,16 +104,39 @@ where
 		return Ok(());
 	}
 
-	log::trace!(
+	let hash = block.header().hash();
+	let number = *block.header().number();
+
+	log::debug!(
 		"Executing Block: {}:{}, version {}",
 		block.header().hash(),
 		block.header().number(),
 		env.client.runtime_version_at(&BlockId::Hash(block.hash())).map_err(|e| format!("{:?}", e))?.spec_version,
 	);
+	let span_events = Arc::new(Mutex::new(SpansAndEvents { spans: Vec::new(), events: Vec::new() }));
+
+	let storage = {
+		let handler = env
+			.tracing_targets
+			.as_ref()
+			.map(|t| TraceHandler::new(&t, number.into(), hash.as_ref().to_vec(), span_events.clone()));
+
+		let _guard = handler.map(tracing::subscriber::set_default);
+
+		let now = std::time::Instant::now();
+		let block = BlockExecutor::new(api, &env.backend, block)?.block_into_storage()?;
+		log::debug!("Took {:?} to execute block", now.elapsed());
+
+		Storage::from(block)
+	};
+	let traces = Arc::try_unwrap(span_events).map_err(|_| TracingError::NoTraceAccess)?.into_inner();
```

Comment:
> https://doc.rust-lang.org/std/sync/struct.Arc.html#method.try_unwrap
Returns the inner value, if the Arc has exactly one strong reference.

So I guess that this relies on that `the storage block` will be executed first increase the `Arc strong count to 2` and then when it's finished it will  drop the `Arc` and decrease the strong count 1 and then you destroy the Mutex because you know that you it's not aliased/mutual exlusion anymore?

This is not super obvious to me and deserves a comment. `Arc::get_mut` and `Arc::try_unwrap` should be used at little as possible IMHO.

Reply:
An alternative is not destroying the Mutex and draining the Vec's inside of it into two new Vector's, which avoids all this unwrapping logic. That introduces two new allocations though, and since this block execution stuff needs to churn through  millions of blocks (with 1000 - 100K loaded at once) I chose to unwrap the mutex (which I _think_ avoids new mem alloc). Archive is using lots of memory as-is with what it's loading, and the mutex is used here in fairly limited scope (famous last words, though).

Another interesting performance question for #205 as to whether this actually helps

Adding a comment explaining it though :+1: